File: packages/xpute/core/src/golden.ts
// @xpute/core/golden.ts

/**
 * The reader of the golden records, as xpute-core/golden.rs is: lines of
 * `path<TAB>value`, the path the keys and indices down to the value joined by
 * dots. A split per line and no parser, so reading is not itself a thing to
 * verify.
 *
 * Which records this side may read is the rule in golden.rs: a record of a
 * value that never leaves the language it was computed in is that language's,
 * and this has no business asserting it. What this is for are the records that
 * cross — the wire's, where both sides have to produce the same bytes from the
 * same input and the record is the only thing that says so.
 *
 * The files live in the Rust crate that reads them with `include_str!`. There
 * is one copy, not one per language, because a second copy is how the two come
 * to disagree without anything failing.
 */

import type { U8Array } from "./abi/array.ts";
import type { f64, u32, u64 } from "./abi/word.ts";
const CRATE = new URL("../../../../crates/xpute/core/golden/", import.meta.url);

export class Golden {
  private readonly map: Map<string, string>;

  private constructor(src: string) {
    this.map = new Map(
      src.split("\n").filter((l) => l.includes("\t")).map((l) => {
        const at = l.indexOf("\t");
        return [l.slice(0, at), l.slice(at + 1)] as const;
      }),
    );
  }

  /** The record at `path` under the crate's `golden/`. */
  static async load(path: string): Promise<Golden> {
    return new Golden(await Deno.readTextFile(new URL(path, CRATE)));
  }

  has(key: string): boolean {
    return this.map.has(key);
  }

  s(key: string): string {
    const v = this.map.get(key);
    if (v === undefined) throw new Error(`golden: no ${key}`);
    return v;
  }

  u32(key: string): u32 {
    return Number(this.s(key));
  }

  u64(key: string): u64 {
    return BigInt(this.s(key));
  }

  /** A double written through its bits, so that -0 and a NaN's payload survive. */
  f64(key: string): f64 {
    const bits = BigInt(this.s(key));
    const view = new DataView(new ArrayBuffer(8));
    view.setBigUint64(0, bits);
    return view.getFloat64(0);
  }

  bool(key: string): boolean {
    const v = this.s(key);
    if (v === "true") return true;
    if (v === "false") return false;
    throw new Error(`golden: ${key} is ${v}, not a boolean`);
  }

  /** Bytes written as hex. */
  bytes(key: string): U8Array {
    const text = this.s(key);
    const out = new Uint8Array(text.length / 2);
    for (let i = 0; i < out.length; i++) out[i] = parseInt(text.slice(i * 2, i * 2 + 2), 16);
    return out;
  }

  /** Calls `f` with `base.0`, `base.1`, … while an entry has that prefix. */
  each(base: string, f: (key: string) => void): void {
    let k = 0;
    for (;; k++) {
      const prefix = base === "" ? `${k}` : `${base}.${k}`;
      const dotted = `${prefix}.`;
      let found = false;
      for (const key of this.map.keys()) {
        if (key === prefix || key.startsWith(dotted)) {
          found = true;
          break;
        }
      }
      if (!found) break;
      f(prefix);
    }
    if (k === 0) throw new Error(`golden: ${base} is empty`);
  }

  /** How many entries `base` has. */
  len(base: string): u32 {
    let n = 0;
    this.each(base, () => n++);
    return n;
  }
}

File: crates/xpute/core/src/golden.rs
// xpute-core/golden.rs

#![cfg(any(test, feature = "golden"))]

//! The reader of the golden records under each crate's `golden/`: lines of
//! `path<TAB>value`, the path the keys and indices down to the value joined
//! by dots. A split per line and no parser, so reading is not itself a thing
//! to verify.
//!
//! A record is an output that was checked once and then frozen. The
//! TypeScript the Rust was ported from wrote them, and what wrote them is
//! gone; the lines a rule below has since moved were re-recorded from the
//! Rust. They are not all worth the same thing, because the two languages
//! are not held to each other everywhere.
//!
//! **Inside a language, each follows its own.** Where a value never leaves the
//! side that computed it, the two are free to differ — Rust rounds to even
//! through `libm` where JavaScript's `Math` rounds half up, and neither is
//! wrong. A record of one of those holds that implementation to what it did,
//! and nothing more: a deliberate change to it is re-recorded, and an
//! accidental one is what the record is there to catch.
//!
//! **Across a protocol, the results must match exactly.** XTP's packets and
//! TLV's bytes are read by the side that did not write them, so a record of
//! one of those is a claim about both languages. It is never re-recorded to
//! make a test pass: a difference there is not a stale baseline, it is the
//! protocol having come apart, and the fix is in whichever side moved.
//!
//! Under `cargo test` only: the crate's own tests, and another crate's
//! through the `golden` feature on its dev-dependency.

use std::collections::HashMap;
use std::panic::{catch_unwind, AssertUnwindSafe};

pub struct Golden {
    map: HashMap<&'static str, &'static str>,
}

impl Golden {
    /// The vector an `include_str!` read.
    pub fn load(src: &'static str) -> Golden {
        Golden {
            map: src.lines().filter_map(|l| l.split_once('\t')).collect(),
        }
    }

    pub fn has(&self, key: &str) -> bool {
        self.map.contains_key(key)
    }

    pub fn s(&self, key: &str) -> &'static str {
        self.map.get(key).unwrap_or_else(|| crate::bug!(ENOENT))
    }

    pub fn u64(&self, key: &str) -> u64 {
        self.s(key).parse().unwrap_or_else(|_| crate::bug!(EINVAL))
    }

    pub fn i64(&self, key: &str) -> i64 {
        self.s(key).parse().unwrap_or_else(|_| crate::bug!(EINVAL))
    }

    pub fn u32(&self, key: &str) -> u32 {
        self.s(key).parse().unwrap_or_else(|_| crate::bug!(EINVAL))
    }

    pub fn i32(&self, key: &str) -> i32 {
        self.s(key).parse().unwrap_or_else(|_| crate::bug!(EINVAL))
    }

    /// A double written through f64_bits.
    pub fn f64(&self, key: &str) -> f64 {
        let text = self.s(key);
        f64::from_bits(u64::from_str_radix(text.trim_start_matches("0x"), 16).unwrap_or_else(|_| crate::bug!(EINVAL)))
    }

    pub fn bool(&self, key: &str) -> bool {
        match self.s(key) {
            "true" => true,
            "false" => false,
            _ => crate::bug!(EINVAL),
        }
    }

    /// Bytes written as hex.
    pub fn bytes(&self, key: &str) -> Vec<u8> {
        let text = self.s(key);
        (0..text.len())
            .step_by(2)
            .map(|i| u8::from_str_radix(&text[i..i + 2], 16).unwrap_or_else(|_| crate::bug!(EINVAL)))
            .collect()
    }

    /// Calls `f` with `base.0`, `base.1`, … while an entry has that prefix.
    pub fn each(&self, base: &str, mut f: impl FnMut(&str)) {
        let mut k = 0;
        loop {
            let prefix = if base.is_empty() { format!("{k}") } else { format!("{base}.{k}") };
            let dotted = format!("{prefix}.");
            if !self.map.keys().any(|key| *key == prefix || key.starts_with(&dotted)) {
                break;
            }
            f(&prefix);
            k += 1;
        }
        crate::ensure!(k > 0, ENOENT);
    }

    /// How many entries `base` has.
    pub fn len(&self, base: &str) -> usize {
        let mut n = 0;
        self.each(base, |_| n += 1);
        n
    }

    pub fn is_empty(&self) -> bool {
        self.map.is_empty()
    }
}

/// A result, or none where the TypeScript threw — "throw" in a vector.
pub fn caught<T>(f: impl FnOnce() -> T) -> Option<T> {
    let hook = std::panic::take_hook();
    std::panic::set_hook(Box::new(|_| {}));
    let out = catch_unwind(AssertUnwindSafe(f)).ok();
    std::panic::set_hook(hook);
    out
}
